### PR TITLE
feat(TDI-42427): Azure Data Lake Storage Gen2

### PR DESCRIPTION
**Pull request by undx**
## What is it about?
https://jira.talendforge.org/browse/TDI-42427

## TODO
- The converter and format packages should be moved to a `connectors-se-common` module to enhance resuability.
- Should merge connection packages to `azure/azurecommon` (BTW `azurecommon` should be renamed `azure-common`).

### CODE REVIEW

File: azure-dls-gen2/src/test/java/org/talend/components/adlsgen2/AdlsGen2TestBase.java [hunk 1]
```diff
@@ -96,13 +96,18 @@ protected void setUp() throws Exception {
...
 
         final MavenDecrypter decrypter = new MavenDecrypter();
-        Server mvnStorage = decrypter.find("azure-dls-gen2.storage");
-        Server mvnAccountSAS = decrypter.find("azure-dls-gen2.sas");
-        Server mvnAccountSharedKey = decrypter.find("azure-dls-gen2.sharedkey");
-        accountName = mvnAccountSAS.getUsername();
-        storageFs = mvnStorage.getUsername();
-        accountKey = mvnAccountSharedKey.getPassword();
-        sas = mvnAccountSAS.getPassword();
+
```
Comment (rmannibucau) [inline azure-dls-gen2/src/test/java/org/talend/components/adlsgen2/AdlsGen2TestBase.java]: there is an extension for that, see https://talend.github.io/component-runtime/main/1.1.11/tutorial-dev-vs-ci-setup.html#_setting_up_credentials, for junit5 it is @WithMavenServers but injections are the same, it is preferred over custom handling (declarative vs programmatic). It supports fallbacks IIRC.

Comment (undx) [reply azure-dls-gen2/src/test/java/org/talend/components/adlsgen2/AdlsGen2TestBase.java]: Ok, Thanks!

File: azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/output/AdlsGen2Output.java [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+    @Service
+    private final AdlsGen2Service service;
+
+    @Service
+    private final I18n i18n;
+
+    private OutputConfiguration configuration;
+
+    private ContentFormatter formatter;
```
Comment (akhabali) [inline azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/output/AdlsGen2Output.java]: can you please removed unused params

File: azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/common/format/csv/CsvFieldDelimiter.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+import lombok.Getter;
+
+@Getter
+public enum CsvFieldDelimiter {
+    SEMICOLON(';'),
+    COMMA(','),
+    TABULATION('\t'),
+    SPACE(' '),
+    OTHER((char) 0);
```
Comment (ypiel-talend) [inline azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/common/format/csv/CsvFieldDelimiter.java]: Is it normal to have a '(char)0' ?

File: azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/common/format/parquet/ParquetIterator.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    public boolean hasNext() {
+        try {
+            current = reader.read();
+            if (current == null) {
+                reader.close();
+                return false;
+            }
+            return true;
+        } catch (IOException e) {
+            log.error("[hasNext] {}", e);
```
Comment (ypiel-talend) [inline azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/common/format/parquet/ParquetIterator.java]: Why a log here and not on other 'throw new Exception' ?

File: azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/output/AdlsGen2Output.java (hunk 2 above)
Comment (undx) [reply azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/output/AdlsGen2Output.java]: Sure.

File: azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/datastore/SharedKeyUtils.java [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+    }
+
+    /**
+     * Constructs a canonicalized string for signing a request.
+     *
+     * @return A canonicalized string.
+     */
+    private String buildStringToSign(URL url, String method, Map<String, String> headers) {
+        String contentLength = getStandardHeaderValue(headers, Constants.HeaderConstants.CONTENT_LENGTH);
+        contentLength = contentLength.equals("") ? Constants.EMPTY_STRING : contentLength;
```
Comment (ypiel-talend) [inline azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/datastore/SharedKeyUtils.java]: Strange... In any cases it changes the value of contentLength. If empty set empty, else set the current value.

File: azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/datastore/Constants.java [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+    public static final String ATTR_IS_DIRECTORY = "isDirectory";
+
+    /**
+     * Private Default Ctor
+     */
+    private Constants() {
+        // Private to prevent construction.
+    }
+
+    public static final class MethodConstants {
```
Comment (ypiel-talend) [inline azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/datastore/Constants.java]: Could have been an enum

File: azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/datastore/Constants.java [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+        public static final String UA_AZURE_DLS_GEN2_DATE = "2019";
+
+        public static final String UA_AZURE_DLS_GEN2_TACOKIT = " tck/";
+
+        public static final String UA_AZURE_DLS_GEN2_TACOKIT_RELEASE = "1.1.9";
+
+        public static final String USER_AGENT_AZURE_DLS_GEN2 = UA_AZURE_DLS_GEN2_PREFIX + UA_AZURE_DLS_GEN2_DATE
+                + UA_AZURE_DLS_GEN2_TACOKIT + UA_AZURE_DLS_GEN2_TACOKIT_RELEASE;
+
+        private HeaderConstants() {
```
Comment (ypiel-talend) [inline azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/datastore/Constants.java]: could have been enum

File: azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/common/format/avro/AvroConfiguration.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+import java.io.Serializable;
+
+import org.talend.sdk.component.api.meta.Documentation;
+
+import lombok.Data;
+
+@Data
+@Documentation("Avro Configuration")
+public class AvroConfiguration implements Serializable {
+}
```
Comment (dmytro-sylaiev) [inline azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/common/format/avro/AvroConfiguration.java]: Why do you need avro, json and parquet configuration classes if they'are empty?

File: azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/input/InputMapper.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+import org.talend.sdk.component.api.input.PartitionSize;
+import org.talend.sdk.component.api.input.Split;
+import org.talend.sdk.component.api.meta.Documentation;
+import org.talend.sdk.component.api.service.Service;
+import org.talend.sdk.component.api.service.record.RecordBuilderFactory;
+
+import static java.util.Collections.singletonList;
+
+@Version(1)
+@Icon(value = Icon.IconType.FILE_DATABASE_O)
```
Comment (ypiel-talend) [inline azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/input/InputMapper.java]: Create a jira to update icon when available in TPD

File: azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/output/AdlsGen2Output.java [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+        try {
+            blobWriter.complete();
+        } catch (Exception e) {
+            log.error("[release]", e);
+            throw new AdlsGen2RuntimeException(e.getMessage());
+        }
+    }
+    //
+    //
+    // @PostConstruct
```
Comment (ypiel-talend) [inline azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/output/AdlsGen2Output.java]: Remove commented code

File: azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/common/format/avro/AvroConfiguration.java (hunk 8 above)
Comment (undx) [reply azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/common/format/avro/AvroConfiguration.java]: If we uprade/enhance configurations later. And the converter builders/formatters have the same kind of signature.

File: azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/runtime/output/CSVBlobWriter.java [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+        if (getBatch().isEmpty()) {
+            return;
+        }
+        byte[] contentBytes = formatter.feedContent(getBatch());
+        uploadContent(contentBytes);
+        getBatch().clear();
+        currentItem.setBlobPath("");
+    }
+
+    // @Override
```
Comment (ypiel-talend) [inline azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/runtime/output/CSVBlobWriter.java]: Remove commented code

File: azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/service/AdlsGen2APIClient.java [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+     */
+    @Request(path = "/")
+    Response<JsonObject> filesystemList( //
+            @Headers() Map<String, String> headers, //
+            @QueryParams(encode = false) Map<String, String> sas, //
+            @Query("resource") String resource, //
+            @Query("timeout") Integer timeout //
+    );
+
+    // GET http://{accountName}.{dnsSuffix}/?resource=account&prefix={prefix}&continuation={continuation}&maxResults
```
Comment (ypiel-talend) [inline azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/service/AdlsGen2APIClient.java]: To remove ? or in javadoc ?

File: azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/service/AdlsGen2APIClient.java [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+            @QueryParams(encode = false) Map<String, String> sas, //
+            @Query("directory") String directory, //
+            @Query("resource") String resource, //
+            @Query("recursive") Boolean recursive, //
+            @Query("continuation") String continuation, //
+            @Query("maxResults") Integer maxResults, //
+            @Query("timeout") Integer timeout //
+    );
+
+    // GET http://{accountName}.{dnsSuffix}/{filesystem}/{path}?timeout={timeout}
```
Comment (ypiel-talend) [inline azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/service/AdlsGen2APIClient.java]: same

File: azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/service/AdlsGen2APIConfigurer.java [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+
+    @Override
+    public void configure(final Connection connection, final ConfigurerConfiguration configuration) {
+        final AdlsGen2Connection conn = configuration.get("connection", AdlsGen2Connection.class);
+        final String auth = configuration.get("auth", String.class);
+        // log.warn("[configure] auth: {}", auth, connection.getHeaders().keySet().toArray());
+        for (String header : connection.getHeaders().keySet()) {
+            log.warn("[configure] {}={}", header, connection.getHeaders().get(header));
+        }
+        // connection //
```
Comment (ypiel-talend) [inline azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/service/AdlsGen2APIConfigurer.java]: commented code to remove ?

File: azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/service/SupportPatch.java [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+import java.lang.reflect.Field;
+import java.lang.reflect.Modifier;
+import java.net.HttpURLConnection;
+import java.util.Arrays;
+import java.util.LinkedHashSet;
+import java.util.Set;
+
+public class SupportPatch {
+
+    public static void main(String... args) throws IOException {
```
Comment (ypiel-talend) [inline azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/service/SupportPatch.java]: To remove

File: azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/datastore/SharedKeyUtils.java (hunk 5 above)
Comment (undx) [reply azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/datastore/SharedKeyUtils.java]: It's needed for the API call, I need to fill a payload (empty or not) to get the content-length set in headers. That's mandatory for some calls to Gen2 API.

File: azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/common/format/csv/CsvFieldDelimiter.java (hunk 3 above)
Comment (undx) [reply azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/common/format/csv/CsvFieldDelimiter.java]: In that case, another property will be used and needed to have a char.

File: azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/datastore/Constants.java (hunk 6 above)
Comment (undx) [reply azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/datastore/Constants.java]: may change in future release.

File: azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/datastore/SharedKeyUtils.java (hunk 5 above)
Comment (ypiel-talend) [reply azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/datastore/SharedKeyUtils.java]: What I don't understand is that you test "" to set "" or current value. Doing nothing is the same no ?
Constants.EMPTY_STRING = "" : https://github.com/Talend/connectors-se/pull/146/files#diff-24f675f6947b38e58c111026d56f9301R76

File: azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/datastore/Constants.java [hunk 16]
```diff
@@ -0,0 +1,304 @@
...
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package org.talend.components.adlsgen2.datastore;
+
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.util.Locale;
+
+public class Constants {
```
Comment (dmytro-sylaiev) [inline azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/datastore/Constants.java]: You have same class in azure-common module..
Also have you thought to have the whole azure-adls2 module inside azure aggregator?

File: azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/datastore/SharedKeyUtils.java (hunk 5 above)
Comment (undx) [reply azure-dls-gen2/src/main/java/org/talend/components/adlsgen2/datastore/SharedKeyUtils.java]: Should have put the constant in the ? expr. but yes.

**Pull request merged**